### DISCUSSION TITLE
Represent files as entities

### DISCUSSION THREAD

**Discussion by Warxcell** [symfony/symfony#42376] [Q&A]
Hello all :) I would like to show off my own kinda bundle (but it can be used also outside Symfony/Doctrine) which takes care of managing user uploads or any generic file.

https://github.com/Warxcell/files

My idea is simple: store metadata of files in database (could be 1 table, multiple table or embeddables) - and don't bother the storage for these. (in the end - any storage is slow, right?). This also gives us ability to easily reference any file from any other entity - by just making association to the file.

My bundle consist of 2 main parts:

Naming strategies = these take care of how are files organized in underlaying storage.
Path resolvers = these take care of generating downloadable url for any file.

Under the hood I use FlySystem (but I consider removing that dependency, got few cases which I can cover very easily - like https://github.com/thephpleague/flysystem/pull/1325).

It has also built-in integration with Liip Imagine.

I will appreciate your feedback. Thank you!

**Comment by javiereguiluz** [comment]
Hi!

There's a popular file upload bundle called [VichUploaderBundle](https://github.com/dustin10/VichUploaderBundle). Could you tell us the main differences between both so we can better decide which one to use? Thanks!

**Reply by Warxcell** [reply]
@javiereguiluz the main difference is the DX (in my opinion my bundle is simplified and lightweight) and the fact that my bundle doesn't have following issue: https://github.com/dustin10/VichUploaderBundle/blob/master/docs/known_issues.md#the-file-is-not-updated-if-there-are-no-other-changes-in-the-entity

Direct comparison: 

VichUploader:

```
<img src="{{ vich_uploader_asset(product, 'image') }}" alt="{{ product.name }}">
```

vs mine:

```
<img src="{{ file_path(product.image) }}" alt="{{ product.name }}">
```

![image](https://user-images.githubusercontent.com/3340882/128313212-9d406af4-bd93-40f5-8f5e-1cabeceb4d39.png)

When I was testing VichUploaded - these properties in the object was very confusing to me - which one should i use ? `image` vs `imageFile` ? Why it's so overcomplicated? Why I have to put a temporary property on my entity and mess around it? I decided to create my own bundle to simplify it. 

```
Next, you have to create the two fields needed for the bundle to work:

create a field (e.g. imageName) which will be stored to the database as a string. This will hold the filename of the uploaded file.
create another field (e.g. imageFile). This will store the UploadedFile object after the form is submitted. This should not be persisted to the database, but you do need to annotate it.
```

Also I don't rely on annotations/attributes - but on interfaces. I prefer the interface

```
interface MutableFile {
   public function setSize(int $size);
   public function setFilename(string $filename);
}
```
vs
```
@Vich\UploadableField(mapping="product_image", fileNameProperty="imageName", size="imageSize")
```

My bundle also allows uploading of any `SplFileInfo`, not only Symfony UploadedFile.

It's lightweight because it have very small requirements:

https://github.com/Warxcell/files/blob/4.0/composer.json#L10-L15

https://github.com/dustin10/VichUploaderBundle/blob/master/composer.json#L17-L30

**Reply by garak** [reply]
Just a note: in VichUploaderBundle you can just use `<img src="{{ vich_uploader_asset(product) }}" alt="{{ product.name }}">`, as long as your entity has a single mapped property (that's true in mostly cases)

**Reply by Warxcell** [reply]
I see. Anyway I didn't liked the magic behind

```
Next, you have to create the two fields needed for the bundle to work:

create a field (e.g. imageName) which will be stored to the database as a string. This will hold the filename of the uploaded file.
create another field (e.g. imageFile). This will store the UploadedFile object after the form is submitted. This should not be persisted to the database, but you do need to annotate it.
```

it's not straightforward IMO
